### PR TITLE
test: enable beartype and jaxtyping everywhere

### PR DESCRIPTION
**Pull request by LeonHilf**
Apparently this wasnt working properly anywhere^^

Now its activated across all packages and including jax typehints. @nicow-elia See the comments for stuff where i was not 100% sure regarding the jax implications

### CODE REVIEW

File: packages/dc_solver_pkg/src/toop_engine_dc_solver/jax/topology_looper.py [hunk 1]
```diff
@@ -509,8 +509,8 @@ def run_solver_inj_bruteforce(
                 0 if disconnections is not None else None,
                 0,
                 0,
-                jax.tree_util.tree_map(lambda _: None, dynamic_information),
```
Comment (LeonHilf) [inline packages/dc_solver_pkg/src/toop_engine_dc_solver/jax/topology_looper.py]: @nicow-elia Here im really unsure.. Background is, that tree_map returns an object of the same kind as the tree you put it. So in this case DynamicInformation with all fields = None -> beartype did not like it. But if its important to have it this way for jax, it can be achieved in another way.

File: packages/topology_optimizer_pkg/src/toop_engine_topology_optimizer/benchmark/benchmark_utils.py [hunk 2]
```diff
@@ -11,8 +11,6 @@
 Note that these methods are simply wrappers to use existing functionality in a sequence.
 """
 
-from __future__ import annotations
```
Comment (LeonHilf) [inline packages/topology_optimizer_pkg/src/toop_engine_topology_optimizer/benchmark/benchmark_utils.py]: @nicow-elia These mess with the jax_typed decorator. For now i removed it and replaced the ForwardRefs by their string
branch_limits: BranchLimits -> branch_limits: "BranchLimits"
With python3.14 this wont be necessary anymore.
See https://github.com/beartype/beartype/discussions/299 aswell

File: .github/copilot-instructions.md [hunk 3]
```diff
@@ -84,11 +83,12 @@ def compute_flows(
 - **Never** put batch-varying data in `StaticInformation` - it forces expensive recompilation.
 - Use `jax_dataclasses.Static` wrapper for pytree fields that shouldn't be traced:
 ```python
-from jax_dataclasses import Static, pytree_dataclass
-@pytree_dataclass
-class MyData:
+from jax_dataclasses import Static
+import equinox as eqx
+
+class MyData(eqx.Module):
```
Comment (LeonHilf) [inline .github/copilot-instructions.md]: @nicow-elia I replaced @pytree_dataclass with eqx.Modules everywhere. This plays nicely with jaxtyping (same creator) and enables all of the runtime checks out of the box. See https://docs.kidger.site/jaxtyping/api/runtime-type-checking/
https://github.com/patrick-kidger/equinox

Comment (nicow-elia) [reply .github/copilot-instructions.md]: lovely

File: packages/dc_solver_pkg/src/toop_engine_dc_solver/jax/topology_looper.py (hunk 1 above)
Comment (nicow-elia) [reply packages/dc_solver_pkg/src/toop_engine_dc_solver/jax/topology_looper.py]: In an earlier version it needed this, but it seems it works fine now? Tests still pass?

File: packages/dc_solver_pkg/src/toop_engine_dc_solver/jax/types.py [hunk 4]
```diff
@@ -541,21 +536,21 @@ class ActionSet:
...
     actions are guaranteed to be in order of substation"""
 
-    unsplit_action_mask: Bool[Array, " n_actions"]
+    unsplit_action_mask: Bool[Array, " n_unsplit_actions"]
     """A mask of unsplit actions, i.e. columns in the actions array that are all False.
     This is used to avoid sampling the unsplit action in the random_topology generator. This is True where
     the unsplit action is stored."""
 
-    reassignment_distance: Int[Array, " n_actions"]
+    reassignment_distance: Int[ArrayLike, " n_reassignment_actions"]
```
Comment (nicow-elia) [inline packages/dc_solver_pkg/src/toop_engine_dc_solver/jax/types.py]: Same here

File: packages/importer_pkg/src/toop_engine_importer/worker/preprocessor.py
Comment (nicow-elia) [inline packages/importer_pkg/src/toop_engine_importer/worker/preprocessor.py]: outdated

File: packages/topology_optimizer_pkg/src/toop_engine_topology_optimizer/ac/scoring_functions.py [hunk 5]
```diff
@@ -61,6 +61,8 @@ def get_early_stopping_contingency_ids(
     for topo, base_case_id in zip(strategy, add_base_case_ids or [None] * len(strategy), strict=True):
         worst_k_contingency_cases = deepcopy(topo.worst_k_contingency_cases)
         if len(worst_k_contingency_cases) == 0:
+            # TODO Does this make sense?
+            # Shouldnt we just continue?
```
Comment (nicow-elia) [inline packages/topology_optimizer_pkg/src/toop_engine_topology_optimizer/ac/scoring_functions.py]: You mean we should raise here? Maybe... But I remember faintly that this did make sense at some point in time

File: packages/topology_optimizer_pkg/src/toop_engine_topology_optimizer/dc/repertoire/discrete_me_repertoire.py [hunk 6]
```diff
@@ -261,7 +260,7 @@ def add_to_repertoire_without_cell_depth(
     new_fitnesses = repertoire.fitnesses.at[batch_of_indices].set(batch_of_fitnesses.squeeze(axis=-1), mode="drop")
     new_descriptors = repertoire.descriptors.at[batch_of_indices].set(batch_of_descriptors, mode="drop")
     if batch_of_extra_scores is None:
-        new_extra_scores = None
+        new_extra_scores = {}
```
[the remark this answers is not in the document]
Comment (nicow-elia) [reply packages/topology_optimizer_pkg/src/toop_engine_topology_optimizer/dc/repertoire/discrete_me_repertoire.py]: In the end it doesn't matter if it's None or empty dict, except for beartype

File: packages/topology_optimizer_pkg/src/toop_engine_topology_optimizer/dc/ga_helpers.py [hunk 7]
```diff
@@ -47,11 +62,11 @@ def state_update(
         """Overwrite the state update to store information for the running means."""
         assert emitter_state is not None
         assert extra_scores is not None
-        return {
-            "total_branch_combis": emitter_state["total_branch_combis"] + extra_scores["n_branch_combis"].astype(int),
-            "total_inj_combis": emitter_state["total_inj_combis"] + extra_scores["n_inj_combis"].astype(int),
-            "total_num_splits": emitter_state["total_num_splits"] + extra_scores["n_split_grids"].astype(int),
-        }
+        return MixingEmitterState(
```
[the remark this answers is not in the document]
Comment (nicow-elia) [reply packages/topology_optimizer_pkg/src/toop_engine_topology_optimizer/dc/ga_helpers.py]: You're right with the assumption that it's always these 3 things, except in Qdax its typed more flexibly. -> solve when removing qdax

**Review by nicow-elia — COMMENTED** [review]
I think my main concern is that jax will trace every loop due to hashing not working anymore, maybe we can introduce a test for that on optimizer level?

File: packages/dc_solver_pkg/src/toop_engine_dc_solver/jax/topology_looper.py (hunk 1 above)
Comment (nicow-elia) [reply packages/dc_solver_pkg/src/toop_engine_dc_solver/jax/topology_looper.py]: Reason was to prevent jax from re-tracing on every iteration

File: packages/topology_optimizer_pkg/src/toop_engine_topology_optimizer/ac/select_strategy.py [hunk 8]
```diff
@@ -217,7 +217,7 @@ def get_repertoire_filter_mask(
 
 
 def get_median_mask(
-    metrics_df: pd.DataFrame, target_metrics: list[MetricType], fitness_col: Optional[Fitness] = "fitness"
+    metrics_df: pd.DataFrame, target_metrics: Sequence[MetricType], fitness_col: Optional[Fitness] = "fitness"
```
[the remark this answers is not in the document]
Comment (LeonHilf) [reply packages/topology_optimizer_pkg/src/toop_engine_topology_optimizer/ac/select_strategy.py]: No, there are two Fitness types. I should probably rename this one

File: packages/importer_pkg/src/toop_engine_importer/network_graph/filter_strategy/empty_bay.py [hunk 9]
```diff
@@ -78,7 +78,7 @@ def get_empty_bay_list(graph: nx.Graph) -> list[list[int]]:
     return empty_asset_bay_lists
 
 
-def get_empty_bay_update_dict(empty_bay_lists: list[list[int]]) -> dict[tuple[int, int], dict[str, WeightValues]]:
+def get_empty_bay_update_dict(empty_bay_lists: list[list[int]]) -> dict[tuple[int, int], dict[str, int]]:
```
[the remark this answers is not in the document]
Comment (LeonHilf) [reply packages/importer_pkg/src/toop_engine_importer/network_graph/filter_strategy/empty_bay.py]: if we would return the WeightValues.low instead of the WeightValues.low.value.
But this leads to a lot of refactoring throughout the code, so i decided to ignore this for now

File: packages/dc_solver_pkg/src/toop_engine_dc_solver/jax/types.py [hunk 10]
```diff
@@ -522,14 +517,14 @@ class ActionSet:
...
-    branch_actions: Bool[Array, " n_actions max_branch_per_sub"]
+    branch_actions: Bool[Array, " n_branch_actions max_branch_per_sub"]
     """A padded out boolean array with the branch assignment for each action in the action set. Each action is a combination
     of a branch and injection assignment. As each action corresponds to a single substation, this holds up to
     max_branch_per_sub booleans and is padded with False if the sub has fewer branches. This holds a True if a branch is on
     bus B and False if it is on bus A. The actions for all substations are concatenated to a large list of n_inj_action
     assignments."""
 
-    inj_actions: Bool[Array, " n_actions max_inj_per_sub"]
+    inj_actions: Bool[Array, " n_inj_actions max_inj_per_sub"]
```
[the remark this answers is not in the document]
Comment (LeonHilf) [reply packages/dc_solver_pkg/src/toop_engine_dc_solver/jax/types.py]: No you are right. fixed

**Review by nicow-elia — APPROVED** [review]
Thank you so much for this work!
